File: Solutions/56_Merge_Intervals/56_Merge_Intervals.rb
#https://leetcode.com/problems/merge-intervals/discuss/804870/Simple-ruby-solution-with-explanation
def merge(a)
    return [] if a == nil || a.size == 0
	# First sort array by first element of each subarray.
    a = a.sort_by{|arr| arr[0]} 
    res = [a[0]]  # Push first subarray in resulting array.
    j = 0  # Keep a pointer for res array.
    for i in 0..a.size-1 do
        # Compare previous and next element using res array.
        # If second element in res subarray is greater than or equal first element in next subarray, it means elements are overlapping.
        # Calculate max by comparing second element in previous and current subarray.
        # Modify res by overriding current subarray.
        if res[j][1] >= a[i][0]
            min = res[j][0]
            max = [a[i][1],res[j][1]].max
            res[j] = [min,max]
        # If second element in res subarray is less than first element in next subarray, it means elements are non overlapping.
        # Push subarray in res and increment pointer for res.
        else
            res << a[i]
            j = j+1
        end
    end
    return res
end

#Third fastest
def merge(intervals)
    return intervals if intervals.length < 2
    
    intervals.sort_by! { |interval| interval[0] }
    result = []
    i = 0
    
    result = [intervals[0]] 
    for i in 1..intervals.length - 1
        if intervals[i][0] <= result[-1][1]
            result[-1][1] = [result[-1][1],intervals[i][1]].max
        else
            result << intervals[i]
        end
    end

    result
end

#Second fastest
def merge(intervals)
    merged = []
    intervals.sort_by! { |interval| interval.first }

    intervals.each do |interval|
      if merged.empty? || merged.last.last < interval.first
        merged << interval
      else
        merged.last[1] = interval.last if interval.last > merged.last.last
      end
    end

    return merged
end

#Fastest
def merge(intervals)
    return intervals if intervals.empty?
    intervals.sort_by!{|interval| interval[0]}
    
    ans = [intervals[0]]
    1.upto(intervals.length-1) do |i|
        if ans[-1][1] >= intervals[i][0]
            ans[-1][1] = [ans[-1][1], intervals[i][1]].max
        else
            ans << intervals[i]
        end
    end
    ans
end

File: Solutions/56_Merge_Intervals/56_Merge_Intervals.js
var merge = function(intervals) {
    if (intervals.length <= 1) return intervals;
    const ans = [];
    intervals.sort((a, b) => a[0] - b[0]);
    
    for (let interval of intervals) {
        if (ans.length == 0 || interval[0] > ans[ans.length-1][1]) {
            ans.push(interval);
        } else {
            ans[ans.length-1][1] = Math.max(ans[ans.length-1][1], interval[1]);
        }
    }
    
    return ans;
};

//Fourth fastest
var merge = function(intervals) {
    intervals.sort((a, b) => {
        return a[0] - b[0];
    });
    let res = [];
    let [st, en] = intervals[0];
    for(let i = 1; i < intervals.length; i++) {
        let [s, e] = intervals[i];
        if(s > en) {
            res.push([st, en]);
            [st, en] = [s, e];
        } else {
            en = Math.max(en, e);
        }
    }
    res.push([st, en]);
    return res;
};

//Third fastest
var merge = function(intervals) {
    if(!intervals || !intervals.length) return []
    
    const res = []
    const n = intervals.length
    
    intervals.sort((a, b) => a[0] - b[0])
    
    let tmp = intervals[0]
    
    for(let i = 1; i < n; i++) {
        if(tmp[1] >= intervals[i][0]) {
            tmp[1] = Math.max(tmp[1], intervals[i][1])
        }
        else {
            res.push(tmp)
            tmp = intervals[i]
        }
    }
    
    res.push(tmp)
    
    return res
};

//Second fastest
var merge = function(intervals) {    
    if (!intervals.length) { return []; }
    
    intervals.sort(function(a, b) { return a[0] - b[0] });
    
    var result  = [];
    var current = intervals.shift();
    var interval;
    
    while (intervals.length) {
        interval = intervals.shift();
        
        if (interval[0] > current[1]) {
            result.push(current);
            current = interval;
        } else if (interval[1] > current[1]) {
            current[1] = interval[1];
        } 
    }
    
    result.push(current);
    
    return result;
};

//Fastest
var merge = function(intervals) {
    // [[1,3],[2,6],[5,10],[15,18]]
    // temp: [1,3], [1,6], [1, 10]
    // newIntervals: [[1,10]]
    if (!intervals.length) { return intervals }
    
    intervals.sort((a,b) => a[0]-b[0])
    let temp = intervals[0]
    const newIntervals = [temp]
    
    for (let i=1; i<intervals.length; i++) {
        if (intervals[i][0] <= temp[1]) {
            temp = [Math.min(temp[0], intervals[i][0]), Math.max(temp[1], intervals[i][1])]
            newIntervals.pop()
            newIntervals.push(temp)
        }
        else {
            temp = intervals[i]
            newIntervals.push(intervals[i])
        }
    }
    
    return newIntervals
};